File: ApiChargeStellarRPC/Mainnet/rpc_passthrough_example.py
#!/usr/bin/env python3
"""
ApiCharge RPC Passthrough Example
=================================

This script demonstrates how to make standard Stellar JSON-RPC calls
via the ApiCharge passthrough route. No fee sponsorship - just direct RPC access.

Requirements:
    pip install stellar-sdk requests

Usage:
    python rpc_passthrough_example.py --secret YOUR_SECRET_KEY

Replace YOUR_SECRET_KEY with your Stellar secret key (starts with 'S').
The account must have USDC balance for token purchase.
"""

import argparse
import base64
import json
import urllib.parse
import requests
from stellar_sdk import Keypair

# =============================================================================
# CONFIGURATION
# =============================================================================

# ApiCharge server URL
SERVER_URL = "https://mainnet.stellar.apicharge.com"
# RPC path (production path)
DEFAULT_RPC_PATH = "/soroban/"

# =============================================================================
# HELPER FUNCTIONS
# =============================================================================

def get_quotes(server_url: str) -> dict:
    """Fetch available route quotes from the server."""
    response = requests.get(f"{server_url}/apicharge/quote", verify=True, timeout=30)
    response.raise_for_status()
    return response.json()


def find_quote_by_route_id(quotes: dict, route_id_substring: str) -> dict | None:
    """Find a quote by partial route ID match."""
    for quote in quotes.get("quotes", []):
        signable = quote.get("signableEntity", {})
        if route_id_substring in signable.get("routeId", ""):
            return quote
    return None


def purchase_access_token(server_url: str, route_quote: dict, keypair: Keypair) -> str:
    """
    Complete the 4-step nanosubscription purchase flow:
    1. Request purchase instruction
    2. Sign the authorization
    3. Submit purchase
    4. Sign the access token
    """
    print("  Step 1: Requesting purchase instruction...")

    purchase_request = {
        "clientPublicKey": keypair.public_key,
        "routeQuote": route_quote
    }

    response = requests.post(
        f"{server_url}/apicharge/nanosubscription/PurchaseInstruction",
        json=purchase_request, verify=True, timeout=60
    )
    if not response.ok:
        print(f"  ERROR: {response.status_code} - {response.text}")
        response.raise_for_status()

    purchase_instruction = response.json()

    print("  Step 2: Signing authorization...")
    auth_to_sign = purchase_instruction.get("authorisationToSign", "")
    auth_bytes = bytes.fromhex(auth_to_sign) if not _is_base64(auth_to_sign) else base64.b64decode(auth_to_sign)
    signed_auth = keypair.sign(auth_bytes)
    purchase_instruction["authorisationToSign"] = base64.b64encode(signed_auth).decode('utf-8')

    print("  Step 3: Purchasing access token...")
    response = requests.post(
        f"{server_url}/apicharge/nanosubscription/Purchase",
        json=purchase_instruction, verify=True, timeout=60
    )
    if not response.ok:
        print(f"  ERROR: {response.status_code} - {response.text}")
        response.raise_for_status()

    access_token = response.json()

    print("  Step 4: Signing access token...")
    signable_entity = access_token.get("signableEntity", {})
    signature_to_sign = signable_entity.get("signature", "")

    if signature_to_sign:
        sig_bytes = base64.b64decode(signature_to_sign)
        token_signature = keypair.sign(sig_bytes)
        access_token["signature"] = base64.b64encode(token_signature).decode('utf-8')

    token_json = json.dumps(access_token, separators=(',', ':'))
    return urllib.parse.quote(token_json, safe='')


def _is_base64(s: str) -> bool:
    try:
        base64.b64decode(s)
        return True
    except:
        return False


# =============================================================================
# MAIN SCRIPT
# =============================================================================

def main():
    parser = argparse.ArgumentParser(description='ApiCharge RPC Passthrough Example')
    parser.add_argument('--secret', '-s', required=True, help='Your Stellar secret key (starts with S)')
    parser.add_argument('--server', default=SERVER_URL, help='ApiCharge server URL')
    parser.add_argument('--path', default=DEFAULT_RPC_PATH, help='RPC path (default: /soroban/)')
    args = parser.parse_args()

    print("=" * 60)
    print("ApiCharge RPC Passthrough Example")
    print("=" * 60)
    print()

    try:
        keypair = Keypair.from_secret(args.secret)
    except Exception as e:
        print(f"ERROR: Invalid secret key: {e}")
        return

    print(f"Account: {keypair.public_key}")
    print(f"Server:  {args.server}")
    print(f"RPC Path: {args.path}")
    print()

    try:
        # Step 1: Fetch quotes
        print("[1/3] Fetching available quotes...")
        quotes = get_quotes(args.server)

        rpc_quote = find_quote_by_route_id(quotes, "stellar-rpc-developer")
        if not rpc_quote:
            print("ERROR: stellar-rpc-developer route not found")
            return

        price = rpc_quote["signableEntity"]["microUnitPrice"] / 1_000_000
        print(f"       RPC route price: ${price:.4f} USDC per call")
        print()

        # Step 2: Purchase access token
        print("[2/3] Purchasing RPC access token...")
        token = purchase_access_token(args.server, rpc_quote, keypair)
        print("       Access token acquired!")
        print()

        # Step 3: Make RPC calls
        print("[3/3] Making RPC calls via passthrough...")
        print()

        # Example 1: getHealth
        print("  >> getHealth")
        response = requests.post(
            f"{args.server}{args.path}",
            headers={"apicharge": token, "Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "id": 1, "method": "getHealth"},
            timeout=30
        )
        result = response.json()
        print(f"     Status: {result.get('result', {}).get('status', 'unknown')}")
        print()

        # Example 2: getLatestLedger
        print("  >> getLatestLedger")
        response = requests.post(
            f"{args.server}{args.path}",
            headers={"apicharge": token, "Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "id": 2, "method": "getLatestLedger"},
            timeout=30
        )
        result = response.json()
        ledger = result.get('result', {})
        print(f"     Sequence: {ledger.get('sequence', 'unknown')}")
        print(f"     Hash: {ledger.get('hash', 'unknown')[:16]}...")
        print()

        # Example 3: getNetwork
        print("  >> getNetwork")
        response = requests.post(
            f"{args.server}{args.path}",
            headers={"apicharge": token, "Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "id": 3, "method": "getNetwork"},
            timeout=30
        )
        result = response.json()
        network = result.get('result', {})
        print(f"     Passphrase: {network.get('passphrase', 'unknown')[:30]}...")
        print()

        print("=" * 60)
        print("SUCCESS! RPC passthrough working correctly")
        print("=" * 60)

    except requests.exceptions.RequestException as e:
        print(f"ERROR: Network error: {e}")
    except Exception as e:
        print(f"ERROR: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    main()

File: ApiChargeStellarRPC/Mainnet/rpc_passthrough_example.js
#!/usr/bin/env node
/**
 * ApiCharge RPC Passthrough Example
 * ==================================
 *
 * This script demonstrates how to make standard Stellar JSON-RPC calls
 * via the ApiCharge passthrough route. No fee sponsorship - just direct RPC access.
 *
 * Requirements:
 *     npm install @stellar/stellar-sdk
 *
 * Usage:
 *     node rpc_passthrough_example.js --secret YOUR_SECRET_KEY
 */

const { Keypair } = require('@stellar/stellar-sdk');

// =============================================================================
// CONFIGURATION
// =============================================================================

const DEFAULT_SERVER_URL = 'https://mainnet.stellar.apicharge.com';
const DEFAULT_RPC_PATH = '/soroban/';  // Production path

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

async function getQuotes(serverUrl) {
    const response = await fetch(`${serverUrl}/apicharge/quote`);
    if (!response.ok) throw new Error(`Failed to fetch quotes: ${response.status}`);
    return response.json();
}

function findQuoteByRouteId(quotes, routeIdSubstring) {
    return quotes.quotes?.find(q =>
        q.signableEntity?.routeId?.includes(routeIdSubstring)
    );
}

async function purchaseAccessToken(serverUrl, routeQuote, keypair) {
    console.log('  Step 1: Requesting purchase instruction...');

    const purchaseRequest = {
        clientPublicKey: keypair.publicKey(),
        routeQuote: routeQuote
    };

    let response = await fetch(`${serverUrl}/apicharge/nanosubscription/PurchaseInstruction`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(purchaseRequest)
    });

    if (!response.ok) {
        const error = await response.text();
        throw new Error(`Purchase instruction failed: ${response.status} - ${error}`);
    }

    const purchaseInstruction = await response.json();

    console.log('  Step 2: Signing authorization...');
    const authToSign = purchaseInstruction.authorisationToSign;
    const authBytes = isBase64(authToSign)
        ? Buffer.from(authToSign, 'base64')
        : Buffer.from(authToSign, 'hex');
    const signedAuth = keypair.sign(authBytes);
    purchaseInstruction.authorisationToSign = signedAuth.toString('base64');

    console.log('  Step 3: Purchasing access token...');
    response = await fetch(`${serverUrl}/apicharge/nanosubscription/Purchase`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(purchaseInstruction)
    });

    if (!response.ok) {
        const error = await response.text();
        throw new Error(`Purchase failed: ${response.status} - ${error}`);
    }

    const accessToken = await response.json();

    console.log('  Step 4: Signing access token...');
    const signatureToSign = accessToken.signableEntity?.signature;
    if (signatureToSign) {
        const sigBytes = Buffer.from(signatureToSign, 'base64');
        const tokenSignature = keypair.sign(sigBytes);
        accessToken.signature = tokenSignature.toString('base64');
    }

    return encodeURIComponent(JSON.stringify(accessToken));
}

function isBase64(str) {
    try {
        return Buffer.from(str, 'base64').toString('base64') === str;
    } catch {
        return false;
    }
}

// =============================================================================
// MAIN SCRIPT
// =============================================================================

async function main() {
    const args = process.argv.slice(2);
    const secretIndex = args.indexOf('--secret');
    const serverIndex = args.indexOf('--server');
    const pathIndex = args.indexOf('--path');
    const secretKey = secretIndex >= 0 ? args[secretIndex + 1] : args.find(a => a.startsWith('S'));
    const SERVER_URL = serverIndex >= 0 ? args[serverIndex + 1] : DEFAULT_SERVER_URL;
    const RPC_PATH = pathIndex >= 0 ? args[pathIndex + 1] : DEFAULT_RPC_PATH;

    if (!secretKey) {
        console.log('Usage: node rpc_passthrough_example.js --secret YOUR_SECRET_KEY [--server URL] [--path /soroban/]');
        console.log('       Your secret key should start with "S"');
        console.log('       --path: RPC path (default: /soroban/)');
        process.exit(1);
    }

    console.log('='.repeat(60));
    console.log('ApiCharge RPC Passthrough Example (JavaScript)');
    console.log('='.repeat(60));
    console.log();

    let keypair;
    try {
        keypair = Keypair.fromSecret(secretKey);
    } catch (e) {
        console.log(`ERROR: Invalid secret key: ${e.message}`);
        process.exit(1);
    }

    console.log(`Account: ${keypair.publicKey()}`);
    console.log(`Server:  ${SERVER_URL}`);
    console.log(`RPC Path: ${RPC_PATH}`);
    console.log();

    try {
        // Step 1: Fetch quotes
        console.log('[1/3] Fetching available quotes...');
        const quotes = await getQuotes(SERVER_URL);

        const rpcQuote = findQuoteByRouteId(quotes, 'stellar-rpc-developer');
        if (!rpcQuote) {
            console.log('ERROR: stellar-rpc-developer route not found');
            process.exit(1);
        }

        const price = rpcQuote.signableEntity.microUnitPrice / 1_000_000;
        console.log(`       RPC route price: $${price.toFixed(4)} USDC per call`);
        console.log();

        // Step 2: Purchase access token
        console.log('[2/3] Purchasing RPC access token...');
        const token = await purchaseAccessToken(SERVER_URL, rpcQuote, keypair);
        console.log('       Access token acquired!');
        console.log();

        // Step 3: Make RPC calls
        console.log('[3/3] Making RPC calls via passthrough...');
        console.log();

        // Example 1: getHealth
        console.log('  >> getHealth');
        let response = await fetch(`${SERVER_URL}${RPC_PATH}`, {
            method: 'POST',
            headers: { 'apicharge': token, 'Content-Type': 'application/json' },
            body: JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'getHealth' })
        });
        let result = await response.json();
        console.log(`     Status: ${result.result?.status || 'unknown'}`);
        console.log();

        // Example 2: getLatestLedger
        console.log('  >> getLatestLedger');
        response = await fetch(`${SERVER_URL}${RPC_PATH}`, {
            method: 'POST',
            headers: { 'apicharge': token, 'Content-Type': 'application/json' },
            body: JSON.stringify({ jsonrpc: '2.0', id: 2, method: 'getLatestLedger' })
        });
        result = await response.json();
        console.log(`     Sequence: ${result.result?.sequence || 'unknown'}`);
        console.log(`     Hash: ${(result.result?.hash || 'unknown').substring(0, 16)}...`);
        console.log();

        // Example 3: getNetwork
        console.log('  >> getNetwork');
        response = await fetch(`${SERVER_URL}${RPC_PATH}`, {
            method: 'POST',
            headers: { 'apicharge': token, 'Content-Type': 'application/json' },
            body: JSON.stringify({ jsonrpc: '2.0', id: 3, method: 'getNetwork' })
        });
        result = await response.json();
        console.log(`     Passphrase: ${(result.result?.passphrase || 'unknown').substring(0, 30)}...`);
        console.log();

        console.log('='.repeat(60));
        console.log('SUCCESS! RPC passthrough working correctly');
        console.log('='.repeat(60));

    } catch (e) {
        console.log(`ERROR: ${e.message}`);
        console.error(e);
        process.exit(1);
    }
}

main();
